### PR TITLE
CategoryFacet component

### PR DESCRIPTION
**Pull request by dms1lva**
[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/HierarchicalFacet/HierarchicalFacet.ts [hunk 1]
```diff
@@ -93,6 +93,10 @@ interface IFlatHierarchy {
  *
  * Selecting the `folder3` value in the facet would only return the `text2.txt` item.
  *
+ * @deprecated This component is exposed for legacy reasons. You should instead use the {@link CategoryFacet} component. It uses a different
```
Comment (amoreauCoveo) [inline src/ui/HierarchicalFacet/HierarchicalFacet.ts]: lol I hate the description that was already there, so I'll edit it all. Here goes:
```
 /**
  * @deprecated This component is exposed for legacy reasons. Instead, use the {@link CategoryFacet} component, which is more performant and easier to use.
  * 
  * The `HierarchicalFacet` component inherits all of its options and behaviors from the [`Facet`]{@link Facet}
  * component, but is meant to be used to render hierarchical values.
  *
  * **Note:**
  * > The `HierarchicalFacet` component does not support the [`customSort`]{@link Facet.options.customSort}
  * > `Facet` option.
  *
  * The `HierarchicalFacet` component can be used to display files in a file system, or categories for items in a
  * hierarchy.
  *
  * This facet requires a group by field with a special format to work correctly.
  *
  * **Example:**
  *
  * You have the following files indexed on a file system:
  * ```
  * c:\
  *   folder1\
  *     text1.txt
  *   folder2\
  *     folder3\
  *       text2.txt
  * ```
  * The `text1.txt` item would have a field with the following format:
  * `c; c|folder1;`
  *
  * The `text2.txt` item would have a field with the following format:
  * `c; c|folder2; c|folder2|folder3;`
  *
  * By default, the `|` character determines the hierarchy (`folder3` inside `folder2` inside `c`).
  *
  * Since both items contain the `c` value, selecting it value in the facet would return both items.
  *
  * Selecting the `folder3` value in the facet would only return the `text2.txt` item.
  * 
  * @notSupportedIn salesforcefree
  */
```

File: sass/_CategoryFacet.scss [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    background-color: $color-blueish-white-grey;
+  }
+}
+
+.CoveoCategoryFacet {
+  &.coveo-hidden {
+    display: none;
+  }
+
+  & > .coveo-facet-values > .coveo-category-facet-value {
```
Comment (olamothe) [inline sass/_CategoryFacet.scss]: I don't like those kind of precise selectors, they are very hard to customize. We should try to make things as flat as possible.

I understand it is "fighting" the base styling from the facet, but could we refactor facet styling or do it more intelligently in any way ?

This goes for this whole file, if we could make it so that "ideally" there is 0 nesting, only 1 rule for each styling we want to apply that would be perfect. I understand there will be exception to that rule, but it should be just that, an exception, and not the norm/standard.

File: src/controllers/CategoryFacetQueryController.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      injectionDepth: this.categoryFacet.options.injectionDepth,
+      maximumNumberOfValues,
+      delimitingCharacter: this.categoryFacet.options.delimitingCharacter
+    };
+    queryBuilder.categoryFacets.push(categoryFacetsRequest);
+    return positionInQuery;
+  }
+
+  public async searchFacetValues(value: string): Promise<IGroupByValue[]> {
+    let lastQuery = { ...this.categoryFacet.queryController.getLastQuery() };
```
Comment (olamothe) [inline src/controllers/CategoryFacetQueryController.ts]: Could be `const`.

File: src/controllers/HistoryController.ts [hunk 4]
```diff
@@ -99,10 +99,12 @@ export class HistoryController extends RootComponent {
     if (this.initialHashChange) {
       this.initialHashChange = false;
       if (hashHasChanged) {
+        // Using replace avoids adding an entry in the History of the browser.
+        // This means that this new URL will become the new initial URL.
         this.window.location.replace(hash);
         this.logger.trace('History hash modified', hash);
       }
-    } else if (this.ignoreNextHashChange) {
+    } else if (hashHasChanged) {
```
Comment (olamothe) [inline src/controllers/HistoryController.ts]: Is this change correct/intended ? Seems like a totally unrelated file. 🤔

File: src/ui/CategoryFacet/CategoryValue.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    this.element = this.categoryFacetTemplates.buildListElement({ value: this.value, count: this.count });
+    this.label = $$(this.element.find('.coveo-category-facet-value-label'));
+    this.collapseArrow = this.categoryFacetTemplates.buildCollapseArrow();
+    this.categoryChildrenValueRenderer = new CategoryChildrenValueRenderer(this.element, categoryFacetTemplates, this, this.categoryFacet);
+    this.labelOnClick = () => {
+      if (!this.passededMaximumDepth()) {
+        this.onLabelClick();
+      }
+    };
+    this.label.one('click', this.labelOnClick);
```
Comment (olamothe) [inline src/ui/CategoryFacet/CategoryValue.ts]: You are not binding `this`, using fat arrow or `bind(this)`..

File: test/ui/CategoryFacet/CategoryFacetSearchTest.ts [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+export function CategoryFacetSearchTest() {
+  describe('CategoryFacetSearch', () => {
+    let categoryFacetMock: CategoryFacet;
+    let categoryFacetSearch: CategoryFacetSearch;
+    let realDebounce;
+    let fakeGroupByValues: IGroupByValue[];
+    beforeEach(() => {
+      realDebounce = _.debounce;
+      spyOn(_, 'debounce').and.callFake(function(func) {
```
Comment (olamothe) [inline test/ui/CategoryFacet/CategoryFacetSearchTest.ts]: What's that for ?

File: test/ui/CategoryFacet/CategoryFacetSearchTest.ts [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+      expect(container.hasClass('coveo-category-facet-search-container')).toBe(true);
+    });
+
+    it('focus moves the focus to the input element', () => {
+      categoryFacetSearch.facetSearchElement = { input: jasmine.createSpyObj('input', ['focus']) } as any;
+      spyOn(categoryFacetSearch.facetSearchElement, 'input');
+
+      categoryFacetSearch.focus();
+
+      expect(categoryFacetSearch.facetSearchElement.input.focus).toHaveBeenCalled();
```
Comment (olamothe) [inline test/ui/CategoryFacet/CategoryFacetSearchTest.ts]: We could add utilities function in those tests to access the elements needed, instead of having to re-type everytime 
`categoryFacetSearch.facetSearchElement.input` or `categoryFacetSearch.facetSearchElement.searchResults`

File: test/ui/CategoryFacet/CategoryFacetSearchTest.ts [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    });
+
+    it('renders values on focus', done => {
+      categoryFacetSearch.build();
+      categoryFacetSearch.focus();
+      setTimeout(() => {
```
Comment (olamothe) [inline test/ui/CategoryFacet/CategoryFacetSearchTest.ts]: I think we could remove all those `setTimeout` in all those tests and use `jasmine.clock().tick()` Easier to read.

File: test/ui/CategoryFacet/CategoryFacetSearchTest.ts [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+          expect($$(pathValues[i]).text()).toEqual('');
+          expect($$(valueCounts[i]).text()).toEqual((i + 1).toString());
+        }
+        done();
+      });
+    });
+
+    it('renders the path correctly', done => {
+      fakeGroupByValues = FakeResults.createFakeGroupByResult('@field', 'a|b|c', 10).values;
+      categoryFacetMock.categoryFacetQueryController.searchFacetValues = () => new Promise(resolve => resolve(fakeGroupByValues));
```
Comment (olamothe) [inline test/ui/CategoryFacet/CategoryFacetSearchTest.ts]: I would add a dedicated utility function to get stuff like `searchWithNoValues()`, `searchWithValues(values: GroupByValues[]), 

Regroup all tests that needs to resolve the same "searchFacetValues" (for example, all tests that check for "empty values" together) inside the same `describe` scope.

File: test/ui/CategoryFacet/CategoryFacetSearchTest.ts [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+    it('sets the correct classes when there is no results', done => {
+      categoryFacetMock.categoryFacetQueryController.searchFacetValues = () => new Promise(resolve => resolve([]));
+      categoryFacetSearch.build();
+
+      categoryFacetSearch.displayNewValues();
+
+      setTimeout(() => {
+        expect($$(categoryFacetMock.element).hasClass('coveo-no-results')).toBe(true);
+        expect($$(categoryFacetSearch.facetSearchElement.search).hasClass('coveo-facet-search-no-results')).toBe(true);
```
Comment (olamothe) [inline test/ui/CategoryFacet/CategoryFacetSearchTest.ts]: Utility function to test for "hasNoResults()" or "hasResults()" etc.

In general, there's a ton of repetitive check throughout all your tests that could be much simpler to read and easier to modify if we need to in the future (eg: If the "coveo-no-results" gets changed to something else)

File: README.md [hunk 11]
```diff
@@ -14,21 +14,23 @@ All resources will be available under `node_modules/coveo-search-ui/bin`. You ca
...
 
 ## Including the Resources from a CDN
 
-Since the April 2017 release, it is possible to access the resources of any specific Coveo JavaScript Search  Framework 
+Since the April 2017 release, it is possible to access the resources of any specific Coveo JavaScript Search Framework
 official release (from version `1.2537` on) through a content delivery network (CDN).
 
-You can simply use an URL such as `https://static.cloud.coveo.com/searchui/v[VERSION]/[PATH_TO_FILE]`, where you 
-replace `[VERSION]` by the actual release version number you wish to use and `[PATH_TO_FILE]` by the path of the file 
+You can simply use an URL such as `https://static.cloud.coveo.com/searchui/v[VERSION]/[PATH_TO_FILE]`, where you
```
[the remark this answers is not in the document]
Comment (amoreauCoveo) [reply README.md]: lol you're right, it is "a URL" since it starts with a /j/ sound. I missed it sorry :P

File: sass/_CategoryFacet.scss (hunk 2 above)
Comment (dms1lva) [reply sass/_CategoryFacet.scss]: I removed this one since it wasn't even used anymore :neutral_face: As for the other ones, I reworked the way the html is built to be able to use simpler selectors.

File: src/controllers/HistoryController.ts (hunk 4 above)
Comment (dms1lva) [reply src/controllers/HistoryController.ts]: Yes it is intented. I stumbled upon it and made a quick change to make a little bit easier to understand.

File: src/rest/CategoryFacetResult.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import { ICategoryFacetValue } from './CategoryFacetValue';
+
+export interface ICategoryFacetResult {
+  notImplemented?: boolean;
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/rest/CategoryFacetResult.ts]: It signifies that the category facet API is not supported on the current search api. Yes it should be optionnal since it will not be returned if it is supported.

File: src/rest/Query.ts [hunk 13]
```diff
@@ -173,6 +174,10 @@ export interface IQuery {
    * This specifies an array of Group By operations that can be performed on the query results to extract facets
    */
   groupBy?: IGroupByRequest[];
+  /**
+   * This specifies an array of request to retrieve facet values for the CategoryFacet component
```
[the remark this answers is not in the document]
Comment (amoreauCoveo) [reply src/rest/Query.ts]: If they're all written like that, yes 😬

File: src/rest/SearchEndpoint.ts [hunk 14]
```diff
@@ -80,7 +80,8 @@ export class SearchEndpoint implements ISearchEndpoint {
         _.extend(
           {
             restUri: 'http://localhost:8100/rest/search',
-            searchAlertsUri: 'http://localhost:8088/rest/search/alerts/'
+            searchAlertsUri: 'http://localhost:8088/rest/search/alerts/',
+            accessToken: '52d806a2-0f64-4390-a3f2-e0f41a4a73ec'
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/rest/SearchEndpoint.ts]: It's the token for the v7 searchuisamples org. I added it so that when we add the local setup argument, it works by default. It looks like the search api changed its local setup to use v8 so I'll just remove this.

File: src/rest/CategoryFacetResult.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (samisayegh) [reply src/rest/CategoryFacetResult.ts]: Hmm, maybe naming the property `notImplementedBySearchApi` would give a clearer explanation of the role of the property?

File: src/ui/CategoryFacet/CategoryFacet.ts [hunk 15]
```diff
@@ -0,0 +1,689 @@
...
+  }
+
+  /**
+   * Reset the facet to its pristine state.
+   */
+  public reset() {
+    this.changeActivePath(this.options.basePath);
+  }
+
+  public disable() {
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/ui/CategoryFacet/CategoryFacet.ts]: Some methods are not documented on purpose.

File: src/ui/CategoryFacet/CategoryFacet.ts [hunk 16]
```diff
@@ -0,0 +1,689 @@
...
+   * If you don't want to specify a value, you can simply enable {@link CategoryFacet.option.debug} and do an empty query.
+   */
+  public async debugValue(value: string) {
+    const queryBuilder = new QueryBuilder();
+    this.categoryFacetQueryController.addDebugGroupBy(queryBuilder, value);
+    const queryResults = await this.queryController.getEndpoint().search(queryBuilder.build());
+    CategoryFacetDebug.analyzeResults(queryResults.groupByResults[0], this.options.delimitingCharacter);
+  }
+
+  public showWaitingAnimation() {
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/ui/CategoryFacet/CategoryFacet.ts]: I judged that users do not these methods.

File: src/ui/CategoryFacet/CategoryFacetBreadcrumb.ts [hunk 17]
```diff
@@ -0,0 +1,31 @@
+import { $$ } from '../../utils/Dom';
+import { SVGDom } from '../../utils/SVGDom';
+import { SVGIcons } from '../../utils/SVGIcons';
+
+export class CategoryFacetBreadcrumb {
+  constructor(
+    private categoryFacetTitle: string,
+    private path: string[],
+    private onClickHandler: (e: MouseEvent) => void,
+    private valueCaption: string,
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/ui/CategoryFacet/CategoryFacetBreadcrumb.ts]: Added a type `CategoryValueDescriptor` that re-groups the name, the count and the path of a value. Also used it when returning values in public methods.

File: src/ui/CategoryFacet/CategoryFacetDebug.ts [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  private addGroupByForEachPathElement(queryBuilder: QueryBuilder, path: string[]) {
+    path.forEach(pathElement => {
+      this.categoryFacet.categoryFacetQueryController.addDebugGroupBy(queryBuilder, pathElement);
+    });
+  }
+
+  /**
+   *  This method expects group by results from a hierarchical field. It will find values with missing parents and values that are
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/ui/CategoryFacet/CategoryFacetDebug.ts]: Since the class is not documented, the method documentation will not be outputted in the reference documentation. This is fine because it is already exposed in the CategoryFacet component.

[the remark this answers is not in the document]
Comment (amoreauCoveo) [reply src/ui/CategoryFacet/CategoryFacetDebug.ts]: Are you sure? I've seen methods being documented alone without their class

[the remark this answers is not in the document]
Comment (dms1lva) [reply src/ui/CategoryFacet/CategoryFacetDebug.ts]: Yes. The FacetQueryController has its methods documented without having the class documented and it does not show up in the reference documentation. Anyway I removed the documentation for this method.

File: src/utils/Dom.ts [hunk 19]
```diff
@@ -31,7 +31,7 @@ export class Dom {
     this.el = el;
   }
 
-  private static handlers: { eventHandle: Function; fn: EventListener }[] = [];
+  private static handlers: WeakMap<(evt: Event, data: any) => void, (e: CustomEvent) => void> = new WeakMap();
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/utils/Dom.ts]: Yes I made sure it works in IE11, also made sure the type definitions correspond to what is supported in IE11.

File: src/ui/CategoryFacet/CategoryFacetSearch.ts [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+    this.facetSearchElement.input.focus();
+    this.handleFacetSearchFocus();
+  }
+
+  public clear() {
+    this.closeSearchInput();
+    this.container && this.container.detach();
+  }
+
+  public handleKeyboardEvent(event: KeyboardEvent) {
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/ui/CategoryFacet/CategoryFacetSearch.ts]: Moved more code to the `FacetSearchElement` class and added a class that handles user input.

File: src/ui/CategoryFacet/CategoryValue.ts (hunk 5 above)
Comment (dms1lva) [reply src/ui/CategoryFacet/CategoryValue.ts]: this.labelOnClick was already a function defined with a fat arrow

File: test/ui/CategoryFacet/CategoryFacetSearchTest.ts (hunk 6 above)
Comment (dms1lva) [reply test/ui/CategoryFacet/CategoryFacetSearchTest.ts]: `jasmine.clock().tick()` doesn't play well with undescore's debounce. I used this to make the debounce call trigger asap

File: test/ui/CategoryFacet/CategoryFacetSearchTest.ts (hunk 8 above)
Comment (dms1lva) [reply test/ui/CategoryFacet/CategoryFacetSearchTest.ts]: As mentionned before, I could not get the jasmine clock to work with underscore debounce, so I used a debounce mock that requires me to use a timeout.

File: src/rest/CategoryFacetResult.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (dms1lva) [reply src/rest/CategoryFacetResult.ts]: That could help, but doesn't the fact that the `notImplemented` field is in the results returned by the SearchApi already indicate that it's not implemented server side?

File: test/controllers/QueryControllerTest.ts [hunk 21]
```diff
@@ -78,6 +78,16 @@ export function QueryControllerTest() {
       }, 10);
     });
 
+    it('should not return undefined or null when getLastQuery is called and there is no last query', done => {
+      test.cmp.executeQuery();
+      setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply test/controllers/QueryControllerTest.ts]: There shouldn't even be an `executeQuery` here, so no `setTimeout` either. Fixed it

File: test/ui/CategoryFacet/CategoryFacetTest.ts [hunk 22]
```diff
@@ -0,0 +1,373 @@
...
+        field: '@field'
+      });
+      test.cmp.activePath = simulateQueryData.query.categoryFacets[0].path;
+    });
+
+    it('when calling getVisibleParentValues returns all the visible parent values', () => {
+      Simulate.query(test.env, simulateQueryData);
+      const visibleParentValues: string[] = test.cmp.getVisibleParentValues();
+      for (let i = 0; i < test.cmp.activePath.length; i++) {
+        expect(visibleParentValues[i]).toEqual(`parent${i}`);
```
[the remark this answers is not in the document]
Comment (dms1lva) [reply test/ui/CategoryFacet/CategoryFacetTest.ts]: I think it's ok like it is right now. This makes sure that parent and children values don't get mixed up.
